File: p404e/sum_of_left_leaves.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumOfLeftLeaves = function (root) {
  let recHelper = function (rt, isLeft) {
    if (rt == null) return 0;
    if (rt.left == null && rt.right == null) {
      return isLeft ? rt.val : 0;
    } else {
      return recHelper(rt.left, true) + recHelper(rt.right, false);
    }
  };
  return recHelper(root, false);
};
// TEST
[
  ['#', 0],
  ['1,#,#', 0],
  ['1,2,#,#,#', 2],
  ['1,#,3,#,#', 0],
  ['1,2,#,#,3,#,#', 2],
  ['3,9,#,#,20,15,#,#,7,#,#', 24],
].forEach((t) => {
  const actual = sumOfLeftLeaves(Tree.deserialize(t[0]));
  console.log('Sum of left leaves of', t[0], '->', actual);
  console.assert(actual === t[1]);
});

File: p404e/sum_of_left_leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from local_packages.binary_tree import TreeNode


class Solution:
    def sumOfLeftLeaves(self, root: TreeNode, isLeft: bool = False) -> int:
        if root is None:
            return 0
        if root.left is None and root.right is None:
            return root.val if isLeft else 0
        else:
            return self.sumOfLeftLeaves(
                root.left, True
            ) + self.sumOfLeftLeaves(root.right, False)


# TESTS
for tree, expected in [
    ("#", 0),
    ("1,2,#,#,#", 2),
    ("1,#,3,#,#", 0),
    ("1,2,#,#,3,#,#", 2),
    ("3,9,#,#,20,15,#,#,7,#,#", 24),
]:
    sol = Solution()
    actual = sol.sumOfLeftLeaves(TreeNode.deserialize(tree))
    print("Sum of left leaves of", tree, "->", actual)
    assert actual == expected
